File: java/graph.java
import java.util.*;

class Node {
    String val;
    Integer weight;
    Node(String val, Integer weight) {
        this.val = val;
        this.weight = weight;
    }
    public String toString() {
        return "{" + this.val + "=" + this.weight + "}";
    }
}

class Graph {
    Map < String, List < Node >> graph = new HashMap < > ();
    public boolean addVertex(String val) {
        if (graph.get(val) == null) {
            graph.put(val, new LinkedList < > ());
            return true;
        }
        return false;
    }
    public boolean addEdge(String vertex1, String vertex2, int weight) {
        if (graph.get(vertex1) == null || graph.get(vertex2) == null || weight < 0) {
            return false;
        }
        Node node2 = new Node(vertex2, weight);
        graph.get(vertex1).add(node2);
        Node node1 = new Node(vertex1, weight);
        graph.get(vertex2).add(node1);
        return true;
    }
    public boolean removeVertex(String val) {
        List < Node > res = graph.remove(val);
        if (res != null) {
            for (Node node: res) {
                Iterator < Node > iterator = graph.get(node.val).iterator();
                while (iterator.hasNext()) {
                    Node item = iterator.next();
                    if (item.val.equals(val)) {
                        iterator.remove();
                    }
                }
            }
        }
        return res != null;
    }
    public boolean removeEdge(String v1, String v2) {
        if (graph.get(v1) == null || graph.get(v2) == null) {
            return false;
        }
        Iterator < Node > iterator;
        iterator = graph.get(v1).iterator();
        while (iterator.hasNext()) {
            Node item = iterator.next();
            if (item.val.equals(v2)) {
                iterator.remove();
            }
        }
        iterator = graph.get(v2).iterator();
        while (iterator.hasNext()) {
            Node item = iterator.next();
            if (item.val.equals(v1)) {
                iterator.remove();
            }
        }
        return true;
    }
    public void bfs(String from) {
        List < Node > list;
        String temp;
        Map < String, Boolean > map = new HashMap < > ();
        Queue < String > queue = new LinkedList < > ();
        queue.add(from);
        while (queue.size() > 0) {
            temp = queue.poll();
            list = graph.get(temp);
            if (map.get(temp) == null) {
                map.put(temp, true);
                System.out.print(temp + "\n");
            }
            for (Node item: list) {
                if (map.get(item.val) == null) {
                    queue.add(item.val);
                }
            }
        }
    }
    public void dfs(String from) {
        List < Node > list;
        String temp;
        Map < String, Boolean > map = new HashMap < > ();
        Stack < String > stack = new Stack < > ();
        stack.push(from);
        while (stack.size() > 0) {
            temp = stack.pop();
            list = graph.get(temp);
            if (map.get(temp) == null) {
                map.put(temp, true);
                System.out.print(temp + "\n");
            }
            for (Node item: list) {
                if (map.get(item.val) == null) {
                    stack.push(item.val);
                }
            }
        }
    }
    public String toString() {
        return this.graph + "";
    }
}

/*
Examples to use thosee methods
       graph.addVertex("Chennai");
       graph.addVertex("Madurai");
       graph.addVertex("Trichy");
       graph.addVertex("Sivagangai");
       graph.addVertex("Madurai");
       graph.addEdge("Chennai","Madurai",500);
       graph.addEdge("Chennai","Trichy",300);
       graph.addEdge("Trichy","Madurai",200);
       graph.addEdge("Sivagangai","Madurai",20);
       graph.bfs("Chennai");
       graph.dfs("Chennai");
*/

File: js/graph.js
class Node {
    constructor(val, weight) {
        this.val = val;
        this.weight = weight;
    }
}

class Graph {
    constructor() {
        this.graph = {};
    }
    addVertex(vertex) {
        if (!this.graph[vertex]) {
            this.graph[vertex] = [];
            return true;
        }
        return false;
    }
    addEdge(v1, v2, weight) {
        if (!this.graph[v1] || !this.graph[v2]) return false;
        let node1 = new Node(v2, weight);
        let node2 = new Node(v1, weight);
        this.graph[v1].push(node1);
        this.graph[v2].push(node2);
        return true;
    }
    removeVertex(v) {
        if (!this.graph[v]) return false;
        this.graph[v].forEach((vertex) => {
            this.graph[vertex.val] = this.graph[vertex.val].filter(
                (item) => item.val != v
            );
        });
        delete this.graph[v];
        return true;
    }
    removeEdge(v1, v2) {
        if (!this.graph[v1] || !this.graph[v2]) return false;
        this.graph[v1] = this.graph[v1].filter((item) => item.val != v2);
        this.graph[v2] = this.graph[v2].filter((item) => item.val != v1);
        return true;
    }

    bfs(from) {
        let map = {};
        let queue = [];
        queue.push(from);
        while (queue.length) {
            let temp = queue.shift();
            if (!map[temp]) {
                console.log(temp);
                map[temp] = true;
            }
            this.graph[temp].forEach((item) => {
                if (!map[item.val]) {
                    queue.push(item.val);
                }
            });
        }
    }
    dfs(from) {
        let map = {};
        let stack = [];
        stack.push(from);
        while (stack.length) {
            let temp = stack.pop();
            if (!map[temp]) {
                console.log(temp);
                map[temp] = true;
            }
            this.graph[temp].forEach((item) => {
                if (!map[item.val]) {
                    stack.push(item.val);
                }
            });
        }
    }
}

//Examples
// let g = new Graph();
// g.addVertex("Chennai");
// g.addVertex("Madurai");
// g.addVertex("Trichy");
// g.addEdge("Chennai", "Madurai", 500);
// g.addEdge("Chennai", "Trichy", 300);
// g.bfs("Chennai");
// g.dfs("Chennai");
